### ISSUE TITLE
Fix the Balancer tests

### ISSUE THREAD

**Issue by artob** [near/nearcore#3665]
I am currently investigating regressions in the [Balancer test suite](https://github.com/artob/balancer-core). The Balancer test suite used to pass, back in September, with zero failures; but that's no longer the case:

<img width="1242" alt="Screen Shot 2020-12-02 at 19 16 45" src="https://user-images.githubusercontent.com/4963/100907549-20533580-34d3-11eb-98ca-01a9e55fd0ea.png">

```
  79 passing (8m)
  8 failing

  1) Contract: BPool
       Extreme weights
         exitswapExternAmountOut should revert:
     AssertionError: Expected to fail with revert, but failed with: Error: value out-of-bounds (argument="tokenAmountOut", value="-112572582263886790000", code=INVALID_ARGUMENT, version=abi/5.0.0-beta.153)
      at fails (node_modules/truffle-assertions/index.js:152:13)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
      at Context.<anonymous> (test/math_extreme_weights.js:301:13)

  2) Contract: TMath
       BMath
         badd throws on overflow:
     AssertionError: Expected to fail with revert, but failed with: Error: Querying call/evm/view failed: wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 69, 82, 82, 95, 65, 68, 68, 95, 79, 86, 69, 82, 70, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))).
{
  "error": "wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 69, 82, 82, 95, 65, 68, 68, 95, 79, 86, 69, 82, 70, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])))",
  "logs": [],
  "block_height": 825,
  "block_hash": "FuoZq56pWoyaeqBkf2tkQMn2dWnJ3YSsa18TaknBrUYG"
}
      at fails (node_modules/truffle-assertions/index.js:152:13)
      at Context.<anonymous> (test/num.js:15:13)

  3) Contract: TMath
       BMath
         bsub throws on underflow:
     AssertionError: Expected to fail with revert, but failed with: Error: Querying call/evm/view failed: wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 17, 69, 82, 82, 95, 83, 85, 66, 95, 85, 78, 68, 69, 82, 70, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))).
{
  "error": "wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 17, 69, 82, 82, 95, 83, 85, 66, 95, 85, 78, 68, 69, 82, 70, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])))",
  "logs": [],
  "block_height": 825,
  "block_hash": "FuoZq56pWoyaeqBkf2tkQMn2dWnJ3YSsa18TaknBrUYG"
}
      at fails (node_modules/truffle-assertions/index.js:152:13)
      at Context.<anonymous> (test/num.js:19:13)

  4) Contract: TMath
       BMath
         bmul throws on overflow:
     AssertionError: Expected to fail with revert, but failed with: Error: Querying call/evm/view failed: wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 69, 82, 82, 95, 77, 85, 76, 95, 79, 86, 69, 82, 70, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))).
{
  "error": "wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 69, 82, 82, 95, 77, 85, 76, 95, 79, 86, 69, 82, 70, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])))",
  "logs": [],
  "block_height": 825,
  "block_hash": "FuoZq56pWoyaeqBkf2tkQMn2dWnJ3YSsa18TaknBrUYG"
}
      at fails (node_modules/truffle-assertions/index.js:152:13)
      at Context.<anonymous> (test/num.js:23:13)

  5) Contract: TMath
       BMath
         bdiv throws on div by 0:
     AssertionError: Expected to fail with revert, but failed with: Error: Querying call/evm/view failed: wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 12, 69, 82, 82, 95, 68, 73, 86, 95, 90, 69, 82, 79, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))).
{
  "error": "wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 12, 69, 82, 82, 95, 68, 73, 86, 95, 90, 69, 82, 79, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])))",
  "logs": [],
  "block_height": 825,
  "block_hash": "FuoZq56pWoyaeqBkf2tkQMn2dWnJ3YSsa18TaknBrUYG"
}
      at fails (node_modules/truffle-assertions/index.js:152:13)
      at Context.<anonymous> (test/num.js:27:13)

  6) Contract: TMath
       BMath
         bpow throws on base outside range:
     AssertionError: Expected to fail with revert, but failed with: Error: Querying call/evm/view failed: wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 21, 69, 82, 82, 95, 66, 80, 79, 87, 95, 66, 65, 83, 69, 95, 84, 79, 79, 95, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))).
{
  "error": "wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 21, 69, 82, 82, 95, 66, 80, 79, 87, 95, 66, 65, 83, 69, 95, 84, 79, 79, 95, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])))",
  "logs": [],
  "block_height": 825,
  "block_hash": "FuoZq56pWoyaeqBkf2tkQMn2dWnJ3YSsa18TaknBrUYG"
}
      at fails (node_modules/truffle-assertions/index.js:152:13)
      at Context.<anonymous> (test/num.js:31:13)

  7) Contract: BPool
       Binding Tokens
         Fails getting final tokens before finalized:
     AssertionError: Expected to fail with revert, but failed with: Error: Querying call/evm/view failed: wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 17, 69, 82, 82, 95, 78, 79, 84, 95, 70, 73, 78, 65, 76, 73, 90, 69, 68, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))).
{
  "error": "wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 17, 69, 82, 82, 95, 78, 79, 84, 95, 70, 73, 78, 65, 76, 73, 90, 69, 68, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])))",
  "logs": [],
  "block_height": 1169,
  "block_hash": "9sFt7dc16TSyQhx6EA62APn1ynJGMnKHyJRzTYetvjWG"
}
      at fails (node_modules/truffle-assertions/index.js:152:13)
      at Context.<anonymous> (test/pool.js:181:13)

  8) Contract: BPool
       User interactions
         Fails calling weights, balances, spot prices on unbound token:
     AssertionError: Expected to fail with revert, but failed with: Error: Querying call/evm/view failed: wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 13, 69, 82, 82, 95, 78, 79, 84, 95, 66, 79, 85, 78, 68, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))).
{
  "error": "wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 13, 69, 82, 82, 95, 78, 79, 84, 95, 66, 79, 85, 78, 68, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])))",
  "logs": [],
  "block_height": 1394,
  "block_hash": "AR733RfVmk2uveopNM9mDhG6vBSTAUhzUQVGEJFQKDSa"
}
      at fails (node_modules/truffle-assertions/index.js:152:13)
      at Context.<anonymous> (test/pool.js:528:13)
```

**Comment by artob** [comment]
All right, replicating @mikedotexe's exact configuration in https://github.com/artob/balancer-core/commit/37c2aac8869798750e85942a5747ff76033ce208, we are down to the same 8 failures that Mike had observed:

<img width="127" alt="Screen Shot 2020-12-02 at 18 49 04" src="https://user-images.githubusercontent.com/4963/100907231-bb97db00-34d2-11eb-8d26-45334a175561.png">

**Comment by mikedotexe** [comment]
Since the tests take quite a while, this may make it easier to test the recurring error of reversion that we see 7 times, just test the `num.js` file:
`./node_modules/.bin/truffle test test/num.js`

**Comment by mikedotexe** [comment]
Interesting, I wonder if the [truffle-assertions](https://www.npmjs.com/package/truffle-assertions) package is actually the focal point of the error not being understood.

**Comment by artob** [comment]
@mikedotexe By instrumenting [near-api-js](https://github.com/near/near-api-js), I've logged the actual JSON-RPC response; it looks like this:

```json
{
  error: 'wasm execution failed with error: FunctionCallError(EvmError(Revert([8, 195, 121, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 69, 82, 82, 95, 65, 68, 68, 95, 79, 86, 69, 82, 70, 76, 79, 87, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])))',
  logs: [],
  block_height: 6333,
  block_hash: '5DucQkGz7HTZriA8VjfPaBSP3Bwn4ihNPQzfMWj5tBzV'
}
```

Given that the error is passed back as just a string, with no error code we can dispatch on, it's no wonder we aren't detecting that the transaction reverted. The quick and dirty fix would be to look for the substring `"FunctionCallError(EvmError(Revert"` in [near-web3-provider](https://github.com/near/near-web3-provider), while we figure out a proper fix that will have to involve changes in [nearcore](https://github.com/near/nearcore) as well for proper error responses. What do you think?

**Comment by mikedotexe** [comment]
Locally I've changed this line:
https://github.com/rkalis/truffle-assertions/blob/master/index.js#L150
to:
`if (errorType && !error.message.toLowerCase().includes(errorType)) {`
which solves the issue of not seeing "revert" because it's "Revert"
but then it's caught on the following `else if` which wants to see the "reason" which we don't have coming in at all, it seems. So that one is likely a nearcore change?

**Comment by artob** [comment]
> Interesting, I wonder if the [truffle-assertions](https://www.npmjs.com/package/truffle-assertions) package is actually the focal point of the error not being understood.

You were ahead of me. I didn't imagine they had such a hacky check for reversion, but indeed, given this:

https://github.com/rkalis/truffle-assertions/blob/master/index.js#L165

https://github.com/rkalis/truffle-assertions/blob/master/index.js#L205-L207

...all we need to do is ensure including that keyword, in lowercase, in our response.

**Comment by artob** [comment]
@mikedotexe Note also how things apparently used to work (this must be the source of the regression):

https://github.com/near/near-web3-provider/blob/1a97cdec86362a54158badee77ba7365873da2de/src/index.js#L712-L720

**Comment by artob** [comment]
All right, the conclusions from @mikedotexe's and my debug session just now: near-web3-provider _does_ include the requisite code for detecting a reverted transaction, here:

https://github.com/near/near-web3-provider/blob/1a97cdec86362a54158badee77ba7365873da2de/src/index.js#L807-L818

It's suboptimal code, but it should be functional. It's attempting to match for the substring `FunctionCallError(EvmError(Revert("`in the returned JSON-RPC error message.

Note the double-quote character in there: the proximate cause of the regression is that our error messages are now coming back in the form of `FunctionCallError(EvmError(Revert([8, 195, 121, 160,`, i.e., with a byte array containing the response data, instead of a string as expected.

So, the quick fix is to just remove that double-quote character, which makes the match succeed and results in the revert being detected.

The proper fix would be to get nearcore to output a structured JSON-RPC error response where we won't need to do brittle string matching in order to detect a reverted transaction.

**Comment by artob** [comment]
After the aforementioned fix to near-web3-provider, we still have the same number of failures, due to a new error. I'm debugging that now...

<img width="1242" alt="Screen Shot 2020-12-02 at 20 54 25" src="https://user-images.githubusercontent.com/4963/100918064-ab86f800-34e0-11eb-9dc1-ce1be8cef7b9.png">

**Comment by artob** [comment]
With @mikedotexe's latest fixes for near-web3-provider (https://github.com/near/near-web3-provider/pull/62), all the reversion issues are now fixed. Now we just have one last and more curious failure left to troubleshoot:

<img width="1242" alt="Screen Shot 2020-12-03 at 09 23 30" src="https://user-images.githubusercontent.com/4963/100977363-8a0e2680-3549-11eb-8b8c-eebe56c07d2b.png">

**Comment by mikedotexe** [comment]
Pretty interesting as I've run the full test probably 10 times as I iterated through the `balancer-core` repo and never saw that. Seems to have gotten more exotic or completely stopped on my end; odd.

**Comment by artob** [comment]
> Pretty interesting as I've run the full test probably 10 times as I iterated through the `balancer-core` repo and never saw that. Seems to have gotten more exotic or completely stopped on my end; odd.

@mikedotexe Curiously, I also don't get this error on every run. It does seem to fail at better than 50/50 odds, though, for me.

Just in case, I rebuilt my testing environment from scratch today, including building the latest `neard` (a8172d15ce5c29fbbd57874ba51438800809db54) and pulling your latest balancer-core changes (https://github.com/ilblackdragon/balancer-core/pull/3), but I'm still seeing the failure.

For faster iteration, it's possible to run just the failing test as follows (takes about 1 minute, on my box):

```bash
./node_modules/.bin/truffle test test/math_extreme_weights.js
```

I've been digging into this, hope to have something to report for our sync call today.

**Comment by mikedotexe** [comment]
Very interesting. This is making me want to try the same thing using `ganache-cli` and see if it's actually an exotic bug there, too or only happens on our tools.

**Comment by artob** [comment]
> Very interesting. This is making me want to try the same thing using `ganache-cli` and see if it's actually an exotic bug there, too or only happens on our tools.

As of the time of writing, I believe this isn't a problem with nearcore, but instead localized to the client libraries, possibly third-party libraries. It looks like it fails in trying to ABI-encode the parameters for the contract call, so we never even hit the RPC endpoint before the failure.

**Comment by artob** [comment]
I've created this minimal standalone script that reproduces the problem with only a dependence on the [@ethersproject/abi](https://www.npmjs.com/package/@ethersproject/abi) package:

```javascript
import { defaultAbiCoder } from "@ethersproject/abi";

// for function exitswapExternAmountOut(uint balanceOut, uint poolTotal, uint tokenAmountOut)
const result = defaultAbiCoder.encode(
  ['uint', 'uint', 'uint'],
  [
    '0x9D68f362923cA176DFfCD30621d50c0dE8b7E525',
    '-102139655687114580000',
    '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff',
  ]
);
console.log(result);
```

**Comment by mikedotexe** [comment]
Interesting that it sometimes passes, I wonder if there's fuzzing somewhere that's feeding different values? Great investigation. I feel like we can timebox this or else honestly just add a section at the bottom of the Balancer README under **Troubleshooting** mentioning this bug.

**Comment by artob** [comment]
I instrumented the `exitswapExternAmountOut` test case, checking whence all variables used to compute contract parameters originate:

```
it('exitswapExternAmountOut should revert', async () => {
    console.log('>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> exitswapExternAmountOut');
    // Call function
    const poolRatioAfterExitFee = 1.1;
    const tokenRatioBeforeSwapFee = poolRatioAfterExitFee ** (1 / daiNorm);
    const tokenAmountOut = currentDaiBalance * (1 - tokenRatioBeforeSwapFee) * (1 - swapFee * (1 - daiNorm));
    try {
        console.log('poolRatioAfterExitFee', poolRatioAfterExitFee); // constant: 1.1
        console.log('daiNorm', daiNorm); // computed in init: Decimal('49').div(Decimal(Decimal('1').add(Decimal('49'))))
        console.log('tokenRatioBeforeSwapFee', tokenRatioBeforeSwapFee); // computed here
        console.log('currentDaiBalance', currentDaiBalance); // MUTABLE VARIABLE!
        console.log('swapFee', swapFee); // constant: 0.001
        console.log('tokenAmountOut', tokenAmountOut); // computed here
        console.log('exitswapExternAmountOut types', typeof DAI, typeof toWei(String(tokenAmountOut)), typeof MAX);
        console.log('exitswapExternAmountOut', DAI, toWei(String(tokenAmountOut)), MAX);
        await truffleAssert.reverts(
            pool.exitswapExternAmountOut(DAI, toWei(String(tokenAmountOut)), MAX),
            'ERR_MAX_OUT_RATIO',
        );
    } catch (e) {
        console.log('!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!');
        console.log(e);
    }
    console.log('<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< exitswapExternAmountOut');
});
```

The output is as follows:

```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> exitswapExternAmountOut
poolRatioAfterExitFee 1.1
daiNorm 0.98
tokenRatioBeforeSwapFee 1.102141698521085
currentDaiBalance 1102.1437413959127689
swapFee 0.001
tokenAmountOut -112.57258226388679
exitswapExternAmountOut types string string string
exitswapExternAmountOut 0x92202F19DAcB6f281fe9A393fF6df9b8C89A3656 -112572582263886790000 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
AssertionError: Expected to fail with revert, but failed with: Error: value out-of-bounds (argument="tokenAmountOut", value="-112572582263886790000", code=INVALID_ARGUMENT, version=abi/5.0.0-beta.153)
    at fails (balancer-core/node_modules/truffle-assertions/index.js:152:13)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at Context.<anonymous> (balancer-core/test/math_extreme_weights.js:312:17) {
  showDiff: false
}
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< exitswapExternAmountOut
```

The only mutable variable input to the test is `currentDaiBalance`, which is modified in previous tests. However, that's not what makes `tokenAmountOut` negative: that comes from `1 - tokenRatioBeforeSwapFee`, which is computed from hard-coded constants.

So, I tentatively conclude that this particular test case is probably actually broken upstream. (I will try it against an Ethereum node to confirm.) We should just disable it, for now, and perhaps file an issue with upstream.

There _are_ a couple of questions I don't have good answers for as yet: If it's entirely broken, why did it ever pass? And why doesn't it fail for you, @mikedotexe? (Perhaps you might locally put in the instrumented test case code from above and paste the output here?)

**Comment by mikedotexe** [comment]
Thanks, Arto. I added your above test (prefixing the name so it's `NEAR exitswapExtern…`) and it passed. 
I ran this like so:

    ./node_modules/.bin/truffle test test/math_extreme_weights.js --network near_local

Results:

```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> exitswapExternAmountOut
poolRatioAfterExitFee 1.1
daiNorm 0.98
tokenRatioBeforeSwapFee 1.102141698521085
currentDaiBalance 1102.1437413959127689
swapFee 0.001
tokenAmountOut -112.57258226388679
exitswapExternAmountOut types string string string
exitswapExternAmountOut 0x8D6Ce9f60D68E31Ee874C53cAaE2e89AA321cA94 -112572582263886790000 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
Receipt: C2QyR8Q9ej96PzDF9L1iGnEDVPj5XYg1TjicNUdiVG3T
	Failure [evm]: Error: {"index":0,"kind":{"EvmError":{"Revert":"08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000114552525f4d41585f4f55545f524154494f000000000000000000000000000000"}}}
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< exitswapExternAmountOut
      ✓ NEAR exitswapExternAmountOut should revert (2544ms)
…
16 passing (54s)
```

**Comment by artob** [comment]
Thanks, @mikedotexe. Curiouser and curiouser: you also have a negative `tokenAmountOut` (which supports the test case being broken upstream), but for some reason the parameter encoding doesn't fail for you (it should, actually). Could you check which version of `@ethersproject/abi` your `package-lock.json` is pulling in? (Mine is using 5.0.0-beta.153.)

**Comment by mikedotexe** [comment]
Same version for me, as well.

**Comment by artob** [comment]
@mikedotexe All right, I've confirmed that after putting in place your `package-lock.json` file and doing `npm install`, all the tests now pass. That means that a differing version of _some_ dependency of [@ethersproject/abi](https://www.npmjs.com/package/@ethersproject/abi) is causing the discrepancy we are seeing.

Presumably I have a newer version with stricter validation, leading to the observed failure. So, that explains that, and I'm satisfied that we've exhausted this investigation. I'll open a bug report with upstream, as this particular test case is certainly broken.

I'd suggest we commit the `package-lock.json` file to our [near/balancer-core](https://github.com/near/balancer-core) fork (in https://github.com/near/balancer-core/pull/1) so that cloning the repository will produce reliable and reproducible results.
